src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -889,18 +889,24 @@ protected boolean asciiOnly(Object string,
     @CoreMethod(names = "bytes", needsBlock = true)
     public abstract static class StringBytesNode extends YieldingCoreMethodNode {
 
-        @Child private BytesNode bytesNode = BytesNode.create();
+        @Child private TruffleString.GetInternalByteArrayNode getByteArrayNode = TruffleString.GetInternalByteArrayNode
+                .create();
 
         @Specialization
         protected RubyArray bytes(Object string, Nil block,
                 @CachedLibrary(limit = "LIBSTRING_CACHE") RubyStringLibrary strings) {
-            final Rope rope = strings.getRope(string);
-            final byte[] bytes = bytesNode.execute(rope);
+            var tstring = strings.getTString(string);
+            var encoding = strings.getEncoding(string).tencoding;
+            var byteArray = getByteArrayNode.execute(tstring, encoding);
+
+            var bytes = byteArray.getArray();
+            int arrayLength = tstring.byteLength(encoding);
+            int offset = byteArray.getOffset();
 
-            final int[] store = new int[bytes.length];
+            final int[] store = new int[arrayLength];
 
-            for (int n = 0; n < store.length; n++) {
-                store[n] = bytes[n] & 0xFF;
+            for (int i = 0; i < arrayLength; i++) {
+                store[i] = bytes[i + offset] & 0xFF;
```

Comment:
I wonder if we should MaterializeNode + ReadByteNode instead of GetInternalByteArrayNode here, WDYT @djoooooe? (considering both PE code and interpreter)
I think that could avoid an extra byte[] copy in case the string is native.

BTW I was thinking there would be a `byte InternalByteArray#get(int index)` but it seems there isn't, maybe there should? Would be convenient in some cases and also an occasion to link to ReadByteNode if that's generally better.

Reply:
I've changed to MaterializeNode + ReadByteNode here.